python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,263 @@
+            # print results.get()
+
+    # Launch jobs stored in our queues. Method is blocked until both queue(s) are empty
+    def waitFinish(self):
+        while len(self.active_jobs) != 0 or not self.runner_queue.empty() or not self.status_queue.empty():
+            with self.thread_lock:
+                self.handleActiveTests()
+
+            # sleep for just a tick or two
+            sleep(0.1)
+
+        self.runner_pool.close()
+        self.runner_pool.join()
+        self.status_pool.close()
+        self.status_pool.join()
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
+    def handleActiveTests(self):
+        # loop through all the active tests and do something
+        for tester_name in self.active_jobs:
+            tester = self.scheduled_name_to_object[tester_name]
+
+            # Handle long running tests
+            self.do_LongRunningTests(tester)
+
+            # Handle timeouts
+            self.do_TimeoutTests(tester)
+
+    def do_LongRunningTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            min_time = self.default_reported_time
+            if tester.specs.isValid('min_reported_time'):
+                min_time = float(tester.getMinReportTime())
+
+            if clock() - tester.getStartTime() > min_time:
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('RUNNING...', tester.bucket_pending)
+                self.harness.handleTestStatus(tester)
+
+    def do_TimeoutTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            if clock() - tester.getStartTime() > float(tester.getMaxTime()):
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('TIMEOUT', tester.bucket_fail)
+                self.harness.handleTestStatus(tester)
+
+    # Status processing (hand a tester's status back to the TestHarness)
+    def statusRunner(self, queue):
+        tester_name = queue.get()
+        tester = self.scheduled_name_to_object[tester_name]
+
+        # check to see if this test is complete
+        if tester.isFinished():
+            # And the running caveat and
+            if tester.getTestName() in self.reported_jobs and tester.didPass():
+                tester.specs.addParam('caveats', ['FINISHED'], "")
+
+            # Print finalized results
+            # TODO: this prints TIMEOUT tests twice. Fix that.
+            self.harness.handleTestStatus(tester)
+
+        else:
+            # place this job back in the runner queue because it was not ready to launch (prereqs)
+            self._runnerGo(tester.getTestName())
+
+            # We should spin here for just a bit; A runner and status thread will rapidly hand off a
+            # job to each other for tests that can't run yet due to a prereq not being ready.
```

Comment:
> each other

Really? I'm sure this is necessary but I don't think this comment explains the reason very well.